homeassistant/components/wake_on_lan/const.py

```diff
@@ -1,3 +1,11 @@
 """Constants for the Wake-On-LAN component."""
 
+from homeassistant.const import Platform
+
 DOMAIN = "wake_on_lan"
+PLATFORMS = [Platform.SWITCH]
```

Comment:
I honestly think this integration should be migrated to become a button platform.

This has been the discussion on why this one hasn't moved to the UI at this point.


../Frenck

Reply:
When defining a `turn_off` action then it's a switch as yes, it sends a magic packet for on but you also then have an `off` action.

I would think perhaps we could do two steps
1. Add the config flow (modify it slightly so it fits point 2)
2. Change so that if a `turn_off` action is set, it will create a switch. If a `turn_off` action is not set it will create a button.

Doing the above in reverse would be complicated as we can't really add a YAML configuration for a button entity e.g. we need the config flow first.
What do you think?

Reply:
The turn off action isn't a responsibility that the wake on lan integration should have. That's up to the user to automate somehow (automation, blueprint, template entity etc). The wake on lan integration should just provide a way for the user to wake a device.

Reply:
For the migration, the first step is probably to create a repair issue with a template switch configuration suggestion based on the config the user has provided for wake on lan.

Reply:
So then the proposed sequence would be something like this
1. Config flow for template switch #121639
2. Deprecate `turn_off` option and add a repair flow to WoL which migrates to a template switch (+ ping sensor) when `turn_off` action is defined
3. Add config flow
4. Deprecate `switch` and add `button`
5. Remove `switch`

Unless we agree to make an exception here we can't add a button until we have a config flow.
I don't think there needs to be any specific waiting time between the steps but obviously step 2 and 4 create some waiting time until we can remove the `SwitchEntity` from WoL.

Reply:
What if you implement the migration of the current YAML configuration by only creating the button platform, ignoring the `turn_off` action? We could create a repair informing the user of the need to create the switch template to replace the previous behavior.

Maybe this is too big a breaking change, but without the exception mentioned above I don't see any other way.

Reply:
For new users it would be nice if we had template blueprints to make it easy to create a wake on lan template switch.

I think all this may need some further discussion before we start working on anything since this integration has many users.

Reply:
Regarding the button entity and migration, note that we already have a wake on lan service so the migration doesn't technically need a button entity.

Reply:
I rather deprecate the service in the move as well to be honest (single change instead of making changes twice).

Reply:
Including the button platform via config entry with the information imported from YAML and creating a repair informing it but keeping the switch active until the user removes it from the config file could be an alternative? After the depreciation period, the switch stops working.

This would be a compromise between importing to the IU without waiting for a depreciation period to pass and also not having an immediate breaking change.

Reply:
The user needs to remove the config from configuration.yaml. We can't do that automatically. We always have a deprecation period. That's not what were discussing.

Reply:
I have now changed this PR into @MartinHjelmare suggestion meaning we have a config flow (without importing anything) which is creating a config entry which loads a button entity.

The YAML including service + switch remains intact as before.

So if you approve this approach then once the template switch is in place we can make a repair flow (start deprecation) which;
1. If there is a `turn_off` action automatically creates a template switch (+ ping sensor which is used for the state if host was provided)
2. If there is no `turn_off` action then it will import and create a button entity instead of switch

the only open question from my side is how to handle point 1 above (ping sensor) if no host was provided as we need something to "calculate" the state for the new template switch. Maybe needs to be provided in the repair flow.

Reply:
If there's no host the template switch will work in optimistic mode, without a value template.

Note that we'll always create a button entity besides the template switch when migrating, since the button entity will be used for the turn on action of the switch.